### ISSUE TITLE
How to convert the xlm_roberta(pytorch)model to onnx?

### ISSUE THREAD

**Issue by laifuchicago** [deepset-ai/FARM#537]
To Author:
(1) If I want to convert pytorch model( xlm-roberta) to onnx , is there any reference? How to set the parameters such as dummy input? (input id, token_type and attention mask ?) The figure is my xlmroberta trained by FARM
[
![ask deepset1](https://user-images.githubusercontent.com/8537280/93066406-8be3ee80-f6ac-11ea-8da0-38125aea813c.PNG)
](url)
The following code is the sample (pytorch to onnx)
####################################################
model_onnx_path = "model.onnx"

The inputs "input_ids", "token_type_ids" and "attention_mask" are torch tensors of shape batch*seq_len
dummy_input = (input_ids, token_type_ids, attention_mask)
input_names = ["input_ids", "token_type_ids", "attention_mask"]
output_names = ["output"]
'''
convert model to onnx
'''
torch.onnx.export(model, dummy_input, model_onnx_path,
input_names = input_names, output_names = output_names, \verbose=False)
#######################################################################################
(2) If we use onnx, will your Haystack also can be retrieved by Elasticsearch?

Thank you
Jonathan Sung

**Comment by Timoeller** [comment]
Hey @laifuchicago how about converting the model to transformers and using their [onnx conversion notebook](https://github.com/huggingface/transformers/blob/master/notebooks/04-onnx-export.ipynb)?

In transformers the conversion works for xlm-roberta. If it is larger than 2GB you have to set `use_external_format=True` and be sure to update to pytorch 1.6.0

I am not sure about the solution you proposed. I see you also use "token_type_ids" which are not used in xlm-roberta. So it most likely will not work.

**Comment by laifuchicago** [comment]
To Author:
Now the question is can I convert my own model(already trained by my own data) or only their model?

Jonathan Sung

**Comment by Timoeller** [comment]
You can convert FARM models to transformer ones and then continue.

We have conversion scripts between transformers and FARM models, your use case is covered [here](https://github.com/deepset-ai/FARM/blob/master/examples/conversion_huggingface_models.py#L36)

**Comment by laifuchicago** [comment]
To Author:
After I transferred , I got this typeerror. Do you have any idea? Thank you.
![ask deepset2](https://user-images.githubusercontent.com/8537280/93178368-b5615080-f766-11ea-8504-68e42553cc6b.PNG)

Jonathan Sung

**Comment by Timoeller** [comment]
I cannot reproduce that error. Did you install latest FARM + dependencies? 
Can you post the full code you used?
The error says it cannot find the vocab file. XLM-R does not have a vocab file but uses a sentencepiece.bpe.model to tokenize. Possibly the model name string does not include "xlm-r"?

**Comment by laifuchicago** [comment]
To Author:
This is the code that I use.
![ask deepset3](https://user-images.githubusercontent.com/8537280/93187935-f3fd0800-f772-11ea-9b9a-f5f9ce2be1e7.PNG)

Jonathan Sung

**Comment by Timoeller** [comment]
It does not load the correct tokenizer.

Check out https://github.com/deepset-ai/FARM/blob/master/farm/modeling/tokenization.py#L73
your model name is missing the "-" symbol.

**Comment by laifuchicago** [comment]
To Author:
This is the model that I trained by my own data. So I changed the name.
Jonathan Sung

**Comment by Timoeller** [comment]
Agreed : ) this very strict string based model loading is not optimal and should be changed towards auto functionality present in transformers. 

Did the conversion work with "xlm-roberta" present in your model name string?

**Comment by laifuchicago** [comment]
To Author:
It already converted to transformer, but how can I convert to onnx? The notebook just use their model. Can I use my model in local?
Jonathan Sung

**Comment by laifuchicago** [comment]
> Hey @laifuchicago how about converting the model to transformers and using their [onnx conversion notebook](https://github.com/huggingface/transformers/blob/master/notebooks/04-onnx-export.ipynb)?
> 
> In transformers the conversion works for xlm-roberta. If it is larger than 2GB you have to set `use_external_format=True` and be sure to update to pytorch 1.6.0
> 
> I am not sure about the solution you proposed. I see you also use "token_type_ids" which are not used in xlm-roberta. So it most likely will not work.

But in the HuggingFace, why they still have token type ids?
![ask deepset6](https://user-images.githubusercontent.com/8537280/93552669-b0d5ab80-f9a3-11ea-85c6-cef8c53bfc22.PNG)
Jonathan Sung

**Comment by Timoeller** [comment]
In general roberta models are not trained on the NSP task and therefore do not need token type IDs.
When you run the transformers conversion from pytorch to onnx you are getting an output that token type ids are unused for xlm-r. Do you agree and also see this output?

If you need to know the details, please ask questions about ONNX conversion and special inputs related to transformers models in huggingface's transformers repository directly.

**Comment by tanaysoni** [comment]
Hi @laifuchicago, with #557,  XLM-RoBERTa models can now be converted to ONNX format for inference.

You can refer to [onnx_question_answering.py](https://github.com/deepset-ai/FARM/blob/master/examples/onnx_question_answering.py) for an example.
